cmake/pcl_find_cuda.cmake

```
@@ -13,16 +13,28 @@ if(CUDA_FOUND)
   enable_language(CUDA)
   set(HAVE_CUDA TRUE)
   
-  execute_process(COMMAND nvcc --list-gpu-code RESULT_VARIABLE EXIT_CODE OUTPUT_VARIABLE OUTPUT_VAL)
-  if(${EXIT_CODE} GREATER 0)
-    message(FATAL_ERROR "Failed to detect gpu arch/code. Exit code is ${EXIT_CODE}")
+  if (CMAKE_CUDA_COMPILER_ID STREQUAL "NVIDIA")
+    if(${CUDA_VERSION_STRING} VERSION_GREATER "11.0")
+      execute_process(COMMAND ${CMAKE_CUDA_COMPILER} --list-gpu-code RESULT_VARIABLE EXIT_CODE OUTPUT_VARIABLE OUTPUT_VAL)
+      #Remove sm_
+      string(REPLACE "sm_" "" OUTPUT_VAL ${OUTPUT_VAL})
+      #Convert to list
+      string(REPLACE "\n" ";" __CUDA_ARCH_BIN ${OUTPUT_VAL})
+      #Remove last empty entry
+      list(REMOVE_AT __CUDA_ARCH_BIN -1)
+    elseif(${CUDA_VERSION_STRING} VERSION_GREATER "10.0")
+      set(__cuda_arch_bin "3.5 3.7 5.0 5.2 5.3 6.0 6.1 6.2 7.0 7.2 7.5 8.0")
+    elseif(${CUDA_VERSION_STRING} VERSION_GREATER "9.0")
+      set(__cuda_arch_bin "3.0 3.2 3.5 3.7 5.0 5.2 5.3 6.0 6.1 6.2 7.0 7.2 7.5")
+    else()
+      set(__cuda_arch_bin "3.0 3.2 3.5 3.7 5.0 5.2 5.3 6.0 6.1 6.2 7.0 7.2")
+    endif()
+  elseif (CMAKE_CUDA_COMPILER_ID STREQUAL "Clang")
+    set(CMAKE_CUDA_ARCHITECTURES ${CUDA_ARCH_BIN})
```

Sorry mault, it should be:
```suggestion
    set(CMAKE_CUDA_ARCHITECTURES "all")
```

---

But that won't work further down?

---

Mhm maybe we need an extra path for Clang?

---

Hm, to me it sounds that clang cuda support is somewhat limited (7-10.1 cuda version). And we don't have any CIs to check it?

Maybe just don't support it and make a FATAL_ERROR?

https://llvm.org/docs/CompileCudaWithLLVM.html

---

Will try it in a fresh VM in the next days if I there is a way how we can test it.

---

Clang CUDA is detected by CMake since 3.18 - and this is the default version for Ubuntu 21.04. So I think we should enable CUDA on the 21.04 build in #4747

---

Played now a bit with it:

My test CMake script:
```CMake
cmake_minimum_required(VERSION 3.10 FATAL_ERROR)

project(Test CXX CUDA)

message(STATUS "CMAKE_CUDA_ARCHITECTURES: ${CMAKE_CUDA_ARCHITECTURES}")
message(STATUS "CMAKE_CUDA_COMPILER_ID: ${CMAKE_CUDA_COMPILER_ID}")
```

#### Ubuntu 21.04: Default installation - NVIDIA CUDA test ####
Dockerfile:
```Docker
FROM "ubuntu:21.04"

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update \
 && apt-get -V install -y \
      build-essential \
      clang \
      cmake \
      nvidia-cuda-toolkit-gcc \
 && rm -rf /var/lib/apt/lists/*
```

Result:
```
-- The CXX compiler identification is Clang 12.0.0
-- The CUDA compiler identification is NVIDIA 11.2.152
```

#### Ubuntu 21.04: Default installation - Clang CUDA test ####
Dockerfile:
```Docker
FROM "ubuntu:21.04"

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update \
 && apt-get -V install -y \
      build-essential \
      clang \
      cmake \
      nvidia-cuda-toolkit-gcc \
 && rm -rf /var/lib/apt/lists/*

ENV CUDACXX=clang
```

Result:
```
-- The CXX compiler identification is Clang 12.0.0
-- The CUDA compiler identification is unknown
...
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - failed
-- Check for working CUDA compiler: /usr/bin/clang
-- Check for working CUDA compiler: /usr/bin/clang - broken
CMake Error at /usr/share/cmake-3.18/Modules/CMakeTestCUDACompiler.cmake:52 (message):
  The CUDA compiler

    "/usr/bin/clang"

  is not able to compile a simple test program.
...
```

#### Ubuntu 21.04: CMake 3.20.2 - Clang CUDA test ####
Dockerfile:
```Docker
FROM "ubuntu:21.04"

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update \
 && apt-get -V install -y \
      build-essential \
      clang \
      nvidia-cuda-toolkit-gcc \
      wget \
 && rm -rf /var/lib/apt/lists/*

RUN wget https://github.com/Kitware/CMake/releases/download/v3.20.2/cmake-3.20.2.tar.gz && \
  tar -zxf cmake-3.20.2.tar.gz && cd cmake-3.20.2 && \
  ./bootstrap -- -DCMAKE_USE_OPENSSL=OFF && \
  make -j$(nproc) install

ENV CUDACXX=clang
```
-- The CXX compiler identification is Clang 12.0.0
-- The CUDA compiler identification is Clang 12.0.0
```

#### Conclusion +CI test ####

The CUDA compiler can be set to Clang from an environment variable, so we should handle the case. Nevertheless to test it on the CI it is a bit difficult due to this [issue](https://gitlab.kitware.com/cmake/cmake/-/issues/21353).

So we need on of the following things:
- CUDA <= 11.0 && >= CMake 3.18
- >=CMake 3.19.1

Sad thing: Ubuntu 21.04 comes with CUDA 11.2.2 and CMake 3.18.4, so a combination where we can't test it.

---

I have removed the case for Clang for now.
Then we can test it, if we get a CI running with Clang for cuda.
Or someone who has interest in getting working with Clang can add it later.